src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs

```diff
@@ -5149,15 +5150,28 @@ protected sealed override ImmutableArray<ISymbol> GetDeclaredSymbolsCore(SyntaxN
                 return this.GetDeclaredSymbols(field, cancellationToken);
             }
 
-            var symbol = GetDeclaredSymbolCore(declaration, cancellationToken);
-            if (symbol != null)
+            // If the type decl has a primary constructor, return that symbol as well.  This is needed so that if the
+            // 'suppression' or 'generated code' attribute is on the primary constructor (i.e. by using `[method:
+            // SuppressMessage(...)]`, it will be found when walking up to the type declaration.
+            if (declaration is TypeDeclarationSyntax typeDeclaration)
             {
-                return ImmutableArray.Create(symbol);
+                var namedType = GetDeclaredSymbol(typeDeclaration, cancellationToken);
+                var primaryConstructor = TryGetSynthesizedPrimaryConstructor(typeDeclaration);
+
+                return primaryConstructor is null
+                    ? ImmutableArray.Create<ISymbol>(namedType)
+                    : ImmutableArray.Create<ISymbol>(namedType, primaryConstructor.GetPublicSymbol());
             }
 
-            return ImmutableArray.Create<ISymbol>();
+            var symbol = GetDeclaredSymbolCore(declaration, cancellationToken);
+            return symbol != null
+                ? ImmutableArray.Create(symbol)
+                : ImmutableArray<ISymbol>.Empty;
         }
 
+        protected virtual SynthesizedPrimaryConstructor TryGetSynthesizedPrimaryConstructor(TypeDeclarationSyntax node)
```

Comment:
>protected virtual SynthesizedPrimaryConstructor TryGetSynthesizedPrimaryConstructor(TypeDeclarationSyntax node) [](http://example.com/codeflow?start=8&length=111)

I am not comfortable with the proposed refactoring around ```TryGetSynthesizedPrimaryConstructor```. Primarily due to the fact that it is virtual with this specific implementation. Either we make it abstract and make sure that each derived class provides real implementation and test that. Or we should consider an alternative: 
- leaving TryGetSynthesizedPrimaryConstructor where is 
- extracting its portion into a helper in this class:
```
        protected SynthesizedPrimaryConstructor TryGetSynthesizedPrimaryConstructor(NamedTypeSymbol type, TypeDeclarationSyntax node)
        {
            var symbol = (type as SourceMemberContainerTypeSymbol)?.PrimaryConstructor;

            if (symbol?.SyntaxRef.SyntaxTree != node.SyntaxTree || symbol.GetSyntax() != node)
            {
                return null;
            }

            return symbol;
        }

```
- Use the helper in both places.

Reply:
sounds good.